### PR TITLE
CEDS-1436 unified notifications model

### PR DESCRIPTION
**Pull request by MaciejRewera**
This PR introduces new Notification model that can handle data from 3 different response formats.
Additionally, the NotificationFactory and various parsers were created, in order to obtain data from all 3 formats. To make things more clean, the strategy pattern was used for response parsers.

### CODE REVIEW

File: app/uk/gov/hmrc/exports/movements/models/notifications/Notification.scala [hunk 1]
```diff
@@ -16,19 +16,20 @@
...
 package uk.gov.hmrc.exports.movements.models.notifications
 
-import java.time.LocalDateTime
+import java.time.Instant
 
 import play.api.libs.json.Json
 
 final case class Notification(
-  dateTimeReceived: LocalDateTime = LocalDateTime.now(),
+  timestampReceived: Instant = Instant.now(),
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/movements/models/notifications/Notification.scala]: 🥇

File: app/uk/gov/hmrc/exports/movements/models/notifications/NotificationData.scala [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.exports.movements.models.notifications
+
+import play.api.libs.json.Json
+
+final case class NotificationData(
+  messageCode: Option[String] = None ,
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/movements/models/notifications/NotificationData.scala]: I think you need to run scalafmt (theres an extra space before the comma)

File: app/uk/gov/hmrc/exports/movements/models/notifications/NotificationData.scala [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  movementReference: Option[String] = None,
+
+  actionCode: Option[String] = None,
+  errorCode: Option[String] = None
+)
+
+object NotificationData {
+  implicit val format = Json.format[NotificationData]
+
+  def empty = NotificationData()
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/movements/models/notifications/NotificationData.scala]: 👍

File: app/uk/gov/hmrc/exports/movements/models/notifications/NotificationPresentationData.scala [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+ */
+
+package uk.gov.hmrc.exports.movements.models.notifications
+
+import java.time.Instant
+
+import play.api.libs.json.Json
+
+final case class NotificationPresentationData(
+  timestampReceived: Instant = Instant.now(),
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/movements/models/notifications/NotificationPresentationData.scala]: 🥇

File: app/uk/gov/hmrc/exports/movements/models/notifications/NotificationPresentationData.scala [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.exports.movements.models.notifications
+
+import java.time.Instant
+
+import play.api.libs.json.Json
+
+final case class NotificationPresentationData(
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/movements/models/notifications/NotificationPresentationData.scala]: I think it's worth sticking a post it on the tech debt board to talk about why `final` is important (if you think it is) then we can make it a "going forward we should do this"

File: app/uk/gov/hmrc/exports/movements/models/notifications/parsers/ControlResponseParser.scala [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    messageCode = stringOption(responseXml \ XmlTags.messageCode),
+    actionCode = stringOption(responseXml \ XmlTags.actionCode),
+    entries = (responseXml \ XmlTags.ucr).map { ucrNode =>
+      Entry(ucrBlock = Some(UcrBlock(ucr = (ucrNode \ XmlTags.ucr).text, ucrType = (ucrNode \ XmlTags.ucrType).text)))
+    },
+    movementReference = stringOption(responseXml \ XmlTags.movementReference),
+    errorCode = stringOption(responseXml \ XmlTags.error \ XmlTags.errorCode)
+  )
+
+  private def stringOption(node: NodeSeq): Option[String] = if (node.text.trim.nonEmpty) Some(node.text) else None
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/movements/models/notifications/parsers/ControlResponseParser.scala]: Can be simplified to `node.text.trim.filter(_.nonEmpty)`

File: app/uk/gov/hmrc/exports/movements/models/notifications/parsers/MovementResponseParser.scala [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+      )
+    }.headOption
+
+    (ucrBlock, goodsItem, entryStatus) match {
+      case (None, Nil, None) => Seq.empty
+      case _                 => Seq(Entry(ucrBlock = ucrBlock, goodsItem = goodsItem, entryStatus = entryStatus))
+    }
+  }
+
+  private def stringOption(node: NodeSeq): Option[String] = if (node.text.trim.nonEmpty) Some(node.text) else None
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/movements/models/notifications/parsers/MovementResponseParser.scala]: Similar here, can also be written as `node.text.trim.filter(_.nonEmpty)`. Up to you which you think is better

File: app/uk/gov/hmrc/exports/movements/models/notifications/parsers/MovementTotalsResponseParser.scala [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        EntryStatus(
+          ics = stringOption(entryStatusNode \ XmlTags.ics),
+          roe = stringOption(entryStatusNode \ XmlTags.roe),
+          soe = stringOption(entryStatusNode \ XmlTags.soe)
+        )
+      }.headOption
+    )
+  }
+
+  private def stringOption(node: NodeSeq): Option[String] = if (node.text.trim.nonEmpty) Some(node.text) else None
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/movements/models/notifications/parsers/MovementTotalsResponseParser.scala]: This is starting to feel like a utility method that could be pulled out?

File: app/uk/gov/hmrc/exports/movements/models/notifications/parsers/ResponseParserContext.scala [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.exports.movements.models.notifications.parsers
+
+case class ResponseParserContext(responseType: String, parser: ResponseParser)
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/movements/models/notifications/parsers/ResponseParserContext.scala]: If we are going to start using `final` we should probably be consistent in using it or not using it

File: test/unit/uk/gov/hmrc/exports/movements/controllers/NotificationControllerSpec2.scala [hunk 10]
```diff
@@ -27,31 +27,29 @@ import play.api.test.FakeRequest
...
-import uk.gov.hmrc.exports.movements.repositories.{NotificationRepository, SubmissionRepository}
-import uk.gov.hmrc.exports.movements.services.NotificationService
-import unit.uk.gov.hmrc.exports.movements.MockMetrics
 import unit.uk.gov.hmrc.exports.movements.base.AuthTestSupport
+import unit.uk.gov.hmrc.exports.movements.base.UnitTestMockBuilder._
 import utils.NotificationTestData._
 
 import scala.concurrent.Future
 
 class NotificationControllerSpec2
```
Comment (ed0906) [inline test/unit/uk/gov/hmrc/exports/movements/controllers/NotificationControllerSpec2.scala]: Appreciate you didnt change this but why is there a 2 at the end?

File: test/unit/uk/gov/hmrc/exports/movements/models/notifications/parsers/NotificationsComparator.scala [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+ */
+
+package unit.uk.gov.hmrc.exports.movements.models.notifications.parsers
+
+import org.scalatest.MustMatchers
+import uk.gov.hmrc.exports.movements.models.notifications.{Notification, NotificationData}
+
+object NotificationsComparator extends MustMatchers {
+
+  def assertEquality(actual: Notification, expected: Notification): Unit = {
```
Comment (ed0906) [inline test/unit/uk/gov/hmrc/exports/movements/models/notifications/parsers/NotificationsComparator.scala]: We could structure this as a matcher, so we can write `actual mustBe expected`

File: app/uk/gov/hmrc/exports/movements/models/notifications/NotificationPresentationData.scala (hunk 5 above)
Comment (MaciejRewera) [reply app/uk/gov/hmrc/exports/movements/models/notifications/NotificationPresentationData.scala]: I'll put a sticker.

File: test/unit/uk/gov/hmrc/exports/movements/controllers/NotificationControllerSpec2.scala (hunk 10 above)
Comment (MaciejRewera) [reply test/unit/uk/gov/hmrc/exports/movements/controllers/NotificationControllerSpec2.scala]: Changed that to be slightly more descriptive why we are using both.

File: app/uk/gov/hmrc/exports/movements/models/notifications/NotificationPresentationData.scala (hunk 5 above)
Comment (ed0906) [reply app/uk/gov/hmrc/exports/movements/models/notifications/NotificationPresentationData.scala]: Also could we get rid of the Data suffix like on the frontend :)

File: app/uk/gov/hmrc/exports/movements/models/notifications/parsers/MovementTotalsResponseParser.scala (hunk 8 above)
Comment (MaciejRewera) [reply app/uk/gov/hmrc/exports/movements/models/notifications/parsers/MovementTotalsResponseParser.scala]: Transformed that into `object StringOption`

File: test/unit/uk/gov/hmrc/exports/movements/models/notifications/parsers/NotificationsComparator.scala (hunk 11 above)
Comment (MaciejRewera) [reply test/unit/uk/gov/hmrc/exports/movements/models/notifications/parsers/NotificationsComparator.scala]: Turned out it is no longer needed. Deleted the object.

File: app/uk/gov/hmrc/exports/movements/models/notifications/NotificationData.scala (hunk 2 above)
Comment (MaciejRewera) [reply app/uk/gov/hmrc/exports/movements/models/notifications/NotificationData.scala]: Done

File: app/uk/gov/hmrc/exports/movements/models/notifications/parsers/StringOption.scala [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.exports.movements.models.notifications.parsers
+
+
+object StringOption {
+
+  def apply(str: String): Option[String] = if (str == null || str.trim.isEmpty) None else Some(str)
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/movements/models/notifications/parsers/StringOption.scala]: Nice, you can still use  `str.trim.filter(_.nonEmpty)` if you prefer
If you're worried about the nulls you can do it as `Option(str).map(_.trim).filter(_.nonEmpty)`

Comment (MaciejRewera) [reply app/uk/gov/hmrc/exports/movements/models/notifications/parsers/StringOption.scala]: Done

**Pull request merged**